Resource automation

rousik opened the pull request:
This script processes responses in the Resource responses and turns them into md files under _resources/. Few things to note:

1. OAuth authentication upon first run is somewhat tedious. The access has not been "approved by google" so you need to click through the security warnings.
Resources are generated into files with naming scheme YYYY-MM-DD-lowercased_title_of_the_resource.md. Spaces replaced with underscores and only first 25 characters of the resource name is retained.

2. Attributes like country, category etc is taken verbatim from the spreadsheet, with the exception of United States being translated into USA
3. There's functionality to use approved_by column (expected to be in the J column), but this is not enabled yet. We expect human reviewers to sanitize the data (esp. category) and then drop their name into that column. This functionality can be enabled by toggling ENABLE_VERIFICATION constant to True.
4. Script still needs to be run manually. I don't think OAuth tokens are suitable for cron-like automation. As discussed on the other thread the volume of Resources is still low so relying on human ops may be appropriate here.

rousik commented:
@mverzilli and @matiasgarciaisaia, would you mind taking a look and checking if this looks good?

mverzilli commented on _scripts/credentials.json:
```diff
@@ -0,0 +1 @@
+{"installed":{"client_id":"557960053095-9ir0ene9v7c825la7trc10e0scblgf6e.apps.googleusercontent.com","project_id":"quickstart-1585151530357","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"Yd5KI3YVjFBI0Zm66JgBWo9p","redirect_uris":["urn:ietf:wg:oauth ...
```
These secrets shouldn't be here, should they?

rousik replied:
Hmm, I'm not sure how can we avoid not having client_id and client_secret here. This tells google which application (and hence which scopes, in this case spreadsheet read access) it should get token for. It seems that in various examples around the web, these two attributes are part of the client-code. You will still need to exchange these credentials for temporary token when running the app. 

I will try to do some more research to see what are the risks of leaking client_secret or whether there are some best-practices around using oauth in code that lives in public repositories.

rousik replied:
Either way, this client_secret is already compromised ;)

matiasgarciaisaia replied:
Can't we have a "Anyone with the link can read" version of the spreadsheet? So we can pull the data without the need for auth. It's data that we _want to publish_ anyway, so no harm in making the spreadsheet itself public.

If that's not an option, then I'd suggest reading this values from environment variables, and (once we decide were to run this from) configure those. I'd start trying to run this from Github Actions (for consistency's sake), but can try alternatives too.

rousik replied:
After looking into this some more I think that having readable by world spreadsheet would eliminate the need for this tedious auth process, one thing we may want to consider here is redacting out the submitter responses.

One thing that comes to mind, and makes the review process just slightly more clunky, would be to have reviewers copy the relevant fields to another world-readable spreadsheet that will be the ground truth for automation. Adds extra step & extra spreadsheet, but eliminates the code complexity. There doesn't seem to be a lot of incoming resources at the moment so this process seems okay to me for now.

What do you think?

rousik replied:
Eliminating the oauth will also make it possible to run this automation periodically and w/o human interaction. OAuth still needs human interaction to fetch the auth token from the client secrets.

mverzilli commented:
Looks great! Just wondering about the credentials file, can you explain why needs to be in the repo?

matiasgarciaisaia commented on _scripts/import_resources.py:
```diff
@@ -0,0 +1,166 @@
+    We will use YYYY-MM-DD of the submission timestamp followed by first 25
+    characters of the resource title with spaces replaced with underscores.
+    """
+   sanitized_name = row[COL_NAME].lower().replace(" ", "_")[:25]
```
I'd replace any non `a-zA-Z0-9` character by a `_` - you don't want to get an exception if a title says "Covid-19: 1/3 of 60+ is at risk".

matiasgarciaisaia commented on _scripts/import_resources.py:
```diff
@@ -0,0 +1,166 @@
+    return (
+        f"---\n"
+        f"{all_attributes}\n"
+        f"\nURL: {row[COL_URL].strip()}\n"
```
We may prefer to use some YAML library to generate the front-matter - just in case there's some escaping to be done.

rousik replied:
So right now, we need to move the contents into the Approved Resources column and this defacto makes this "trusted" input, but I'm happy to add sanitizer in the code. I'm not sure what's the right sanitizer to use for markdown fragments though.

rousik commented:
PTAL. I have added "Approved Resources" tab to the original spreadsheet and published that tab to web. This means that we can fetch that as CSV w/o needing any authentication. This makes the import script much nicer and simpler. This also gives us ability to sanitize the resource texts as we wish, which means that less sanitization is needed in the code.

I have also reverted changes to _resources to make it simple.

It should be possible to run this script in a cron-like fashion.

rousik commented:
The procedure for absorbing and approving new Resources should be documented somewhere, probably at https://github.com/flattenthecurve/guide/wiki/Processes. We can link this with the slack message to #resources channel to make it smooth.

rousik commented:
@mverzilli and @matiasgarciaisaia , should be good to go now.
